tests/test_type_caster_bare_interface.cpp

```diff
@@ -0,0 +1,215 @@
+// Systematically exercises the detail::type_caster<> interface. This is going a step in the
+// direction of an integration test, to ensure multiple components of pybind11 work together
+// correctly. It is also useful to show the type_caster<> interface virtually clutter-free.
+
+#include "pybind11_tests.h"
+
+#include <memory>
+
+namespace pybind11_tests {
+namespace type_caster_bare_interface {
+
+struct mpty {};
+
+// clang-format off
+
+mpty        rtrn_valu() { mpty obj; return obj; }
+mpty&&      rtrn_rref() { static mpty obj; return std::move(obj); }
+mpty const& rtrn_cref() { static mpty obj; return obj; }
+mpty&       rtrn_mref() { static mpty obj; return obj; }
+mpty const* rtrn_cptr() { return new mpty; }
+mpty*       rtrn_mptr() { return new mpty; }
+
+const char* pass_valu(mpty)        { return "load_valu"; }
+const char* pass_rref(mpty&&)      { return "load_rref"; }
+const char* pass_cref(mpty const&) { return "load_cref"; }
+const char* pass_mref(mpty&)       { return "load_mref"; }
+const char* pass_cptr(mpty const*) { return "load_cptr"; }
+const char* pass_mptr(mpty*)       { return "load_mptr"; }
+
+std::shared_ptr<mpty>       rtrn_shmp() { return std::shared_ptr<mpty      >(new mpty); }
+std::shared_ptr<mpty const> rtrn_shcp() { return std::shared_ptr<mpty const>(new mpty); }
+
+const char* pass_shmp(std::shared_ptr<mpty>)       { return "load_shmp"; }
+const char* pass_shcp(std::shared_ptr<mpty const>) { return "load_shcp"; }
+
+std::unique_ptr<mpty>       rtrn_uqmp() { return std::unique_ptr<mpty      >(new mpty); }
+std::unique_ptr<mpty const> rtrn_uqcp() { return std::unique_ptr<mpty const>(new mpty); }
+
+const char* pass_uqmp(std::unique_ptr<mpty>)       { return "load_uqmp"; }
```

Comment:
Uuumpf, this takes me 2 second per function to parse to know what it's saying. Do you feel strongly about keeping 4 letters? Personally, I find `pass_unique_ptr`, `pass_const_ref`, `pass_rvalue_ref`, `return_...`, ... easier to directly read, even though things then don't nicely line out anymore.

But your effort on writing tests, so your final say.